### PR TITLE
T5 support

### PR DESCRIPTION
**Pull request by davidkaczer**
# What does this PR do?

This PR adds support for pretrained LongT5 encoder-decoder models from HuggingFace.

## General Changes
* Add HuggingFacePretrainedEncoderDecoderModel class for loading LongT5
* Add span masking collator ported from [official Google implementation](https://github.com/huggingface/transformers/blob/main/examples/flax/language-modeling/run_t5_mlm_flax.py)
* Add test for span masking collator
* Add example training configs for LongT5
* Add logic to distinguish between decoder-only and encoder-decoder models when passing inputs

## Breaking Changes
* None intended

## Checklist before submitting final PR
- [x] My PR is minimal and addresses one issue in isolation
- [x] I have merged the latest version of the target branch into this feature branch
- [x] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [x] I have run a sample config for model training
- [ ] I have checked that all tests run through (`python tests/tests.py`) (apparently some tests were failing already in upstream? also didn't test multi-GPU)
- [ ] I have updated the internal changelog (`CHANGELOG_DEV.md`)

### CODE REVIEW

File: config_files/training/config_lorem_ipsum_span_masking_T5.yaml [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+collate_fn:  
+  component_key: collate_fn
+  variant_key: span_masking_collator_wrapper
+  config:
+    sample_key: ${settings.referencing_keys.sample_key}
+    target_key: ${settings.referencing_keys.target_key}
+    noise_density: 0.3
+    mean_noise_span_length: 3
+    tokenizer:
```
Comment (le1nux) [inline config_files/training/config_lorem_ipsum_span_masking_T5.yaml]: need to check if we really need the tokenizer dependency

File: config_files/training/config_lorem_ipsum_span_masking_T5.yaml [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+optimizer:  
+  component_key: optimizer
+  variant_key: adam_w
+  config:
+    lr: 0.0001
+    betas: [0.9, 0.95]
+    eps: 1e-8
+    weight_decay: 1e-1
+    # need to fix in huggingface_model.py to same spec as in gpt2_model.py
+    weight_decay_groups_excluded: []
```
Comment (le1nux) [inline config_files/training/config_lorem_ipsum_span_masking_T5.yaml]: Need to check if this has to be filled

File: examples/getting_started/tokenizer_t5/config.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (le1nux) [inline examples/getting_started/tokenizer_t5/config.json]: Maybe move this outside of getting_started?

File: src/modalities/models/huggingface/collator.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            mask_indices (np.ndarray[bool]): binary mask of tokens to mask out
+
+        Returns:
+            np.ndarray[int]: array of sentinel ids at starting point of spans to be masked out
+        """
+        start_indices = mask_indices - np.roll(mask_indices, 1, axis=-1) * mask_indices
+        start_indices[:, 0] = mask_indices[:, 0]
+
+        sentinel_ids = np.where(start_indices != 0, np.cumsum(start_indices, axis=-1), start_indices)
+        sentinel_ids = np.where(sentinel_ids != 0, (self.tokenizer.vocab_size - sentinel_ids), 0)
```
Comment (le1nux) [inline src/modalities/models/huggingface/collator.py]: we might want to replace the `tokenizer` parameter by the two parameters `eos_token_id` and `vocab_size`

File: src/modalities/models/huggingface/collator.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+            np.ndarray[int]: input ids with masked out spans replaced by single sentinel token
+        """
+        batch_size = input_ids.shape[0]
+
+        input_ids_full = np.where(sentinel_ids != 0, sentinel_ids, input_ids)
+        # input_ids tokens and sentinel tokens are >= 0, tokens < 0 are
+        # masked tokens coming after sentinel tokens and should be removed
+        input_ids = input_ids_full[input_ids_full >= 0].reshape((batch_size, -1))
+        input_ids = np.concatenate(
+            [input_ids, np.full((batch_size, 1), self.tokenizer.tokenizer.eos_token_id)], axis=-1, dtype=np.int32
```
Comment (le1nux) [inline src/modalities/models/huggingface/collator.py]: Need to verify that we actually need the `eos_token_id` here. I would assume that the pbin file already comes with the eos_token for every document.

File: src/modalities/models/model.py [hunk 6]
```diff
@@ -72,6 +72,12 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
 
 def model_predict_batch(model: nn.Module, batch: DatasetBatch) -> InferenceResultBatch:
-    forward_result = model.forward(batch.samples)
+    # Ideally we would do this by checking isinstance(model, HuggingFacePretrainedEncoderDecoderModel),
+    # but HuggingFacePretrainedEncoderDecoderModel depends on NNModel from this file
+    # so this creates a circular dependency
+    if hasattr(model, "huggingface_model") and hasattr(model.huggingface_model, "encoder"):
+        forward_result = model.forward(inputs=batch.samples, targets=batch.targets)
+    else:
+        forward_result = model.forward(inputs=batch.samples)
```
Comment (le1nux) [inline src/modalities/models/model.py]: Can we get rid of this check and have the huggingface model also implement the NNModelIF? We could have a wrapper for the HuggingFace model in this case.

File: examples/getting_started/tokenizer_t5/config.json (hunk 3 above)
Comment (davidkaczer) [reply examples/getting_started/tokenizer_t5/config.json]: Moved to data/

File: src/modalities/models/huggingface/collator.py (hunk 5 above)
Comment (davidkaczer) [reply src/modalities/models/huggingface/collator.py]: Turns out this adds an EOS token at the end of every batch, not document, which seems clearly bad. So removed it

File: config_files/training/config_lorem_ipsum_span_masking_T5.yaml (hunk 2 above)
Comment (davidkaczer) [reply config_files/training/config_lorem_ipsum_span_masking_T5.yaml]: Fixed, logic in the model is already there, just changed the config

File: src/modalities/models/huggingface/collator.py (hunk 4 above)
Comment (davidkaczer) [reply src/modalities/models/huggingface/collator.py]: Done. eos_token_id is no longer needed

File: config_files/training/config_lorem_ipsum_span_masking_T5.yaml (hunk 1 above)
Comment (davidkaczer) [reply config_files/training/config_lorem_ipsum_span_masking_T5.yaml]: Removed, now only pass vocab_size (int)

File: src/modalities/models/model.py (hunk 6 above)
Comment (davidkaczer) [reply src/modalities/models/model.py]: Do you mean NNModel? There is no class NNModelIF.

I am not convinced about adding yet another wrapper class. Can you elaborate on how you would approach this?

IMO we should either
- refactor so that model_predict_batch is a method of NNModel that its subclass HuggingFacePretrainedEncoderDecoderModel inherits and redefines (preferred)
- move model_predict_batch to models.utils and then check type using an import (less preferred because it doesn't create a circular import now but creates the risk of one in the future) 
- leave as is (ugly but effective)

Comment (davidkaczer) [reply src/modalities/models/model.py]: I briefly tried 
- refactor so that model_predict_batch is a method of NNModel that its subclass HuggingFacePretrainedEncoderDecoderModel inherits and redefines (preferred)

and there is some strange behavior. The current implementation as a separate function somehow automatically transfers the batch tensors from CPU to GPU, and changing to NNModel method breaks this. Probably the FSDP wrapper is responsible.

Comment (le1nux) [reply src/modalities/models/model.py]: I just remebered, that we already have an implementation for it. 
https://github.com/Modalities/modalities/blob/main/src/modalities/models/huggingface/huggingface_model.py#L43

I think it would work out-of-the-box with a minor refactoring maybe. 

There are two things which I think we should consider: 
1. In the current branch version, we have low-level if else cases in a high-level routine (`model_predict_batch`). That's quite ugly.
2. `model_predict_batch` was introduced such that the models themselves don't need to know what a batch is. They only have their weights and a forward routine based on an input dict (`Dict[str, torch.Tensor]`). Huggingface tends to place lots of logic into the models, such as targets, loss calculation etc. From our perspective a model must not have any knowledge regarding these things. 

Are we currently also calculating the loss outside of T5? If so, it should possible to easily integrate `HuggingFacePretrainedModel`.

Comment (davidkaczer) [reply src/modalities/models/model.py]: Changed the type check to this:
```if model.model_type.value in [LongT5Model, LongT5ForConditionalGeneration]:```
using the `model_type` attribute  from https://github.com/Modalities/modalities/blob/main/src/modalities/models/huggingface/huggingface_model.py#L43

Is this what you had in mind?

**Comment by davidkaczer** [comment]
Also changed `HuggingFacePretrainedEncoderDecoderModel` to be a child of `HuggingFacePretrainedModel` instead of its sibling. This simplifies potential future type checks and avoids duplicating methods needlessly.
